src/main/kotlin/tanvd/grazi/ide/ui/GraziRulesTree.kt

```diff
@@ -0,0 +1,279 @@
+        val expansionMonitor = TreeExpansionMonitor.install(tree)
+
+        override fun filter() {
+            if (!filter.isNullOrBlank()) expansionMonitor.freeze()
+
+            filter(filter)
+
+            val expandedPaths = TreeUtil.collectExpandedPaths(tree)
+            (tree.model as DefaultTreeModel).reload()
+            TreeUtil.restoreExpandedPaths(tree, expandedPaths)
+
+            SwingUtilities.invokeLater {
+                tree.setSelectionRow(0)
+            }
+
+            TreeUtil.expandAll(tree)
+            if (filter.isNullOrBlank()) {
+                TreeUtil.collapseAll(tree, 0)
+                expansionMonitor.restore()
+            }
+        }
+    }
+
+    init {
+        tree.selectionModel.addTreeSelectionListener { e ->
+            when (val meta = (e?.path?.lastPathComponent as DefaultMutableTreeNode).userObject) {
+                is RuleWithLang -> selectionListener(meta.rule)
+                is Category -> selectionListener(meta)
+                is Lang -> selectionListener(meta)
+            }
+        }
+
+        tree.addCheckboxTreeListener(object : CheckboxTreeListener {
+            override fun mouseDoubleClicked(node: CheckedTreeNode) {
+                // do nothing
+            }
+
+            override fun beforeNodeStateChanged(node: CheckedTreeNode) {
+                // do nothing
+            }
+
+            override fun nodeStateChanged(node: CheckedTreeNode) {
+                val meta = node.userObject
+                if (meta is RuleWithLang) {
+                    meta.state = node.isChecked
+                    if (meta.enabled == meta.state) {
+                        state.remove(meta.rule.id)
+                    } else {
+                        state[meta.rule.id] = meta
+                    }
+                }
+            }
+        })
+
+        val scrollPane = ScrollPaneFactory.createScrollPane(tree)
+        val filterPanel = JPanel(BorderLayout())
+        filterPanel.add(filter, BorderLayout.CENTER)
+        filterPanel.border = JBUI.Borders.emptyBottom(2)
+
+        val group = DefaultActionGroup()
+        val actionManager = CommonActionsManager.getInstance()
+        val treeExpander = DefaultTreeExpander(tree)
+        group.add(actionManager.createExpandAllAction(treeExpander, tree))
+        group.add(actionManager.createCollapseAllAction(treeExpander, tree))
+        filterPanel.add(ActionManager.getInstance().createActionToolbar("GraziRulesTree", group, true).component, BorderLayout.WEST)
+
+        panel.add(filterPanel, BorderLayout.NORTH)
+        panel.add(scrollPane, BorderLayout.CENTER)
+
+        filter.reset()
+    }
+
+    private fun getNodeText(node: CheckedTreeNode): String {
+        return when (val meta = node.userObject) {
+            is RuleWithLang -> meta.rule.description
+            is Category -> meta.name
+            is Lang -> meta.displayName
+            else -> "???"
```

Comment:
Throw error if you see unexpected result

Reply:
It's not unexpected, changed on empty string